pkg/db/table.go

```diff
@@ -0,0 +1,285 @@
+			}
+			if len(fk.Cols) == 0 {
+				// TODO: In MySQL, this case will report a parse error.
+				return nil, infoschema.ErrCannotAddForeign
+			}
+			tbInfo.ForeignKeys = append(tbInfo.ForeignKeys, &fk)
+			continue
+		}
+		if constr.Tp == ast.ConstraintPrimaryKey {
+			var col *table.Column
+			for _, key := range constr.Keys {
+				col = table.FindCol(cols, key.Column.Name.O)
+				if col == nil {
+					return nil, errors.Errorf("key column %s doesn't exist in table", key.Column.Name)
+				}
+				// Virtual columns cannot be used in primary key.
+				if col.IsGenerated() && !col.GeneratedStored {
+					return nil, errors.Errorf("Defining a virtual generated column as primary key")
+				}
+			}
+			if len(constr.Keys) == 1 {
+				switch col.Tp {
+				case mysql.TypeLong, mysql.TypeLonglong,
+					mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24:
+					tbInfo.PKIsHandle = true
+				}
+			}
+		}
+		// build index info.
+		idxInfo, err := buildIndexInfo(tbInfo, model.NewCIStr(constr.Name), constr.Keys, model.StatePublic)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		//check if the index is primary or uniqiue.
+		switch constr.Tp {
+		case ast.ConstraintPrimaryKey:
+			idxInfo.Primary = true
+			idxInfo.Unique = true
+			idxInfo.Name = model.NewCIStr(mysql.PrimaryKeyName)
+		case ast.ConstraintUniq, ast.ConstraintUniqKey, ast.ConstraintUniqIndex:
+			idxInfo.Unique = true
+		}
+		tbInfo.Indices = append(tbInfo.Indices, idxInfo)
+	}
+
+	return
+}
+
+func buildIndexInfo(tblInfo *model.TableInfo, indexName model.CIStr, idxColNames []*ast.IndexColName, state model.SchemaState) (*model.IndexInfo, error) {
+	idxColumns, err := buildIndexColumns(tblInfo.Columns, idxColNames)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Create index info.
+	idxInfo := &model.IndexInfo{
+		Name:    indexName,
+		Columns: idxColumns,
+		State:   state,
+	}
+	return idxInfo, nil
+}
+
+func buildIndexColumns(columns []*model.ColumnInfo, idxColNames []*ast.IndexColName) ([]*model.IndexColumn, error) {
+	idxColumns := make([]*model.IndexColumn, 0, len(idxColNames))
+
+	for _, ic := range idxColNames {
+		col := findCol(columns, ic.Column.Name.O)
+
+		idxColumns = append(idxColumns, &model.IndexColumn{
+			Name:   col.Name,
+			Offset: col.Offset,
+			Length: ic.Length,
+		})
+	}
+
+	return idxColumns, nil
+}
+
+func allocateColumnID(tblInfo *model.TableInfo) int64 {
+	tblInfo.MaxColumnID++
+	return tblInfo.MaxColumnID
+}
+
+// findCol finds column in cols by name.
+func findCol(cols []*model.ColumnInfo, name string) *model.ColumnInfo {
+	// column name don't distinguish capital and small letter
+	name = strings.ToLower(name)
+	for _, col := range cols {
+		if col.Name.L == name {
+			return col
+		}
+	}
+
+	return nil
+}
+
+func setImplicitColumn(table *model.TableInfo) {
+	newColumn := &model.ColumnInfo{
+		ID:   ImplicitColID,
+		Name: model.NewCIStr(ImplicitColName),
+	}
+	newColumn.Tp = mysql.TypeInt24
```

Comment:
TypeInt24 for ImplicitColID?

Reply:
int is ok, no matter int64 or int32, just used for judge type